File: business/src/main/java/com/kevinguanchedarias/owgejava/business/AllianceBo.java
package com.kevinguanchedarias.owgejava.business;

import com.kevinguanchedarias.owgejava.dto.AllianceDto;
import com.kevinguanchedarias.owgejava.entity.Alliance;
import com.kevinguanchedarias.owgejava.entity.AllianceJoinRequest;
import com.kevinguanchedarias.owgejava.entity.UserStorage;
import com.kevinguanchedarias.owgejava.enumerations.AuditActionEnum;
import com.kevinguanchedarias.owgejava.exception.ProgrammingException;
import com.kevinguanchedarias.owgejava.exception.SgtBackendInvalidInputException;
import com.kevinguanchedarias.owgejava.repository.AllianceJoinRequestRepository;
import com.kevinguanchedarias.owgejava.repository.AllianceRepository;
import com.kevinguanchedarias.owgejava.repository.UserStorageRepository;
import com.kevinguanchedarias.owgejava.util.SpringRepositoryUtil;
import lombok.AllArgsConstructor;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.io.Serial;

/**
 * @author Kevin Guanche Darias <kevin@kevinguanchedarias.com>
 * @since 0.7.0
 */
@AllArgsConstructor
@Service
public class AllianceBo implements WithNameBo<Integer, Alliance, AllianceDto> {
    @Serial
    private static final long serialVersionUID = 2632768998010477053L;

    private final AllianceRepository repository;
    private final ConfigurationBo configurationBo;
    private final UserStorageRepository userStorageRepository;
    private final AuditBo auditBo;
    private final AllianceJoinRequestRepository allianceJoinRequestRepository;

    /**
     * @author Kevin Guanche Darias <kevin@kevinguanchedarias.com>
     * @since 0.7.0
     */
    @Override
    public JpaRepository<Alliance, Integer> getRepository() {
        return repository;
    }

    /*
     * (non-Javadoc)
     *
     * @see com.kevinguanchedarias.owgejava.business.BaseBo#getDtoClass()
     */
    @Override
    public Class<AllianceDto> getDtoClass() {
        return AllianceDto.class;
    }

    /**
     * Deletes an alliance <br>
     * <b>NOTICE: </b> Handles unsetting all users
     *
     * @author Kevin Guanche Darias <kevin@kevinguanchedarias.com>
     * @since 0.7.0
     */
    @Transactional
    public void delete(Alliance alliance) {
        defineAllianceByAllianceId(alliance.getId());
        repository.delete(alliance);
    }

    /**
     * Saves an alliance to the database <br>
     * <b>NOTICE:</b> Also sets the user alliance of the owner to the <i>newly</i>
     * created alliance
     *
     * @param invokerId User requesting the save
     * @throws ProgrammingException When the owner is null
     * @author Kevin Guanche Darias <kevin@kevinguanchedarias.com>
     * @since 0.7.0
     */
    @Transactional
    public Alliance save(Alliance alliance, Integer invokerId) {
        if (Boolean.parseBoolean(configurationBo.findOrSetDefault("DISABLED_FEATURE_ALLIANCE", "FALSE").getValue())) {
            throw new SgtBackendInvalidInputException(
                    "You can't not create an alliance, while the idea is nice, it's not possible");
        }
        Alliance retVal;
        if (alliance.getId() == null) {
            var creator = SpringRepositoryUtil.findByIdOrDie(userStorageRepository, invokerId);
            if (creator.getAlliance() != null) {
                throw new SgtBackendInvalidInputException("You already have an alliance, leave it first");
            }
            alliance.setOwner(creator);
            retVal = repository.save(alliance);
            retVal.getOwner().setAlliance(retVal);
            auditBo.doAudit(AuditActionEnum.JOIN_ALLIANCE);
            userStorageRepository.save(retVal.getOwner());
        } else {
            var storedAlliance = findById(alliance.getId());
            checkInvokerIsOwner(storedAlliance, invokerId);
            storedAlliance.setName(alliance.getName());
            storedAlliance.setDescription(alliance.getDescription());
            retVal = repository.save(storedAlliance);
        }
        return retVal;
    }

    /**
     * @return True if:
     * <ul>
     *   <li>The <i>source</i> user has not an alliance</li>
     *   <li>The <i>target</i>> user has not an alliance</li>
     *   <li>The source alliance is different than the target alliance</li>
     *   </ul>
     * @author Kevin Guanche Darias <kevin@kevinguanchedarias.com>
     */
    public boolean areEnemies(UserStorage source, UserStorage target) {
        return !source.getId().equals(target.getId()) && (source.getAlliance() == null || target.getAlliance() == null
                || !source.getAlliance().getId().equals(target.getAlliance().getId()));
    }

    /**
     * Request the entrance in an alliance
     *
     * @return Persisted user
     * @throws SgtBackendInvalidInputException When you already belong to an
     *                                         alliance
     * @author Kevin Guanche Darias <kevin@kevinguanchedarias.com>
     * @since 0.7.0
     */
    public AllianceJoinRequest requestJoin(Integer allianceId, Integer invokerId) {
        var alliance = findByIdOrDie(allianceId);
        var user = SpringRepositoryUtil.findByIdOrDie(userStorageRepository, invokerId);
        if (user.getAlliance() != null) {
            throw new SgtBackendInvalidInputException("You are already in an alliance, nice try!");
        }
        var retVal = new AllianceJoinRequest();
        retVal.setAlliance(alliance);
        retVal.setUser(user);
        return allianceJoinRequestRepository.save(AllianceJoinRequest.builder()
                .alliance(alliance)
                .user(user)
                .build()
        );
    }

    @Transactional
    public void acceptJoin(Integer joinRequestId, Number invoker) {
        var request = SpringRepositoryUtil.findByIdOrDie(allianceJoinRequestRepository, joinRequestId);
        checkInvokerIsOwner(request.getAlliance(), invoker);
        checkIsLimitReached(request);
        if (request.getUser().getAlliance() == null) {
            var alliance = request.getAlliance();
            request.getUser().setAlliance(alliance);
            alliance.getUsers()
                    .forEach(user -> auditBo.nonRequestAudit(AuditActionEnum.USER_INTERACTION, "JOIN_ALLIANCE", request.getUser(), user.getId()));
            auditBo.doAudit(AuditActionEnum.ACCEPT_JOIN_ALLIANCE, null, request.getUser().getId());
            userStorageRepository.save(request.getUser());
            allianceJoinRequestRepository.deleteByUser(request.getUser());
        } else {
            allianceJoinRequestRepository.delete(request);
        }
    }

    /**
     * Rejects the join request
     *
     * @author Kevin Guanche Darias <kevin@kevinguanchedarias.com>
     * @since 0.7.0
     */
    public void rejectJoin(Integer joinRequestId, Number invoker) {
        var request = SpringRepositoryUtil.findByIdOrDie(allianceJoinRequestRepository, joinRequestId);
        checkInvokerIsOwner(request.getAlliance(), invoker);
        allianceJoinRequestRepository.delete(request);
    }

    /**
     * Leaves an alliance
     *
     * @author Kevin Guanche Darias <kevin@kevinguanchedarias.com>
     * @since 0.7.0
     */
    @Transactional
    public void leave(Integer userId) {
        var userRef = userStorageRepository.getReferenceById(userId);
        if (isOwnerOfAnAlliance(userId)) {
            throw new SgtBackendInvalidInputException("You can't leave your own alliance");
        }
        userRef.setAlliance(null);
        userStorageRepository.save(userRef);
    }

    /**
     * Deletes the alliance associated with the user
     *
     * @throws SgtBackendInvalidInputException When user is not the owner of the
     *                                         alliance
     * @author Kevin Guanche Darias <kevin@kevinguanchedarias.com>
     * @since 0.7.0
     */
    @Transactional
    public void deleteByUser(UserStorage transientUser) {
        var user = SpringRepositoryUtil.findByIdOrDie(userStorageRepository, transientUser.getId());
        var alliance = user.getAlliance();
        if (alliance == null) {
            throw new SgtBackendInvalidInputException("You don't have any alliance");
        }
        checkInvokerIsOwner(alliance, user);
        delete(alliance);
    }

    /**
     * Checks if the user is the owner of the request
     *
     * @author Kevin Guanche Darias <kevin@kevinguanchedarias.com>
     * @since 0.7.0
     */
    public void checkInvokerIsOwner(Alliance storedAlliance, UserStorage user) {
        checkInvokerIsOwner(storedAlliance, user.getId());
    }

    /**
     * Checks if the alliance is of user property
     *
     * @throws SgtBackendInvalidInputException When user is not the owner of the
     *                                         alliance
     * @author Kevin Guanche Darias <kevin@kevinguanchedarias.com>
     * @since 0.7.0
     */
    public void checkInvokerIsOwner(Alliance storedAlliance, Number userId) {
        if (!storedAlliance.getOwner().getId().equals(userId)) {
            throw new SgtBackendInvalidInputException(
                    "You are NOT the owner of that alliance, try hacking the owner account");
        }
    }

    /**
     * Returns true if user has an alliance
     *
     * @author Kevin Guanche Darias <kevin@kevinguanchedarias.com>
     * @since 0.7.0
     */
    public boolean isOwnerOfAnAlliance(Number userId) {
        return repository.findOneByOwnerId(userId) != null;
    }

    private void checkIsLimitReached(AllianceJoinRequest request) {
        long userCount = userStorageRepository.count();
        float allowedPercentage = Integer
                .parseInt(configurationBo.findOrSetDefault("ALLIANCE_MAX_SIZE_PERCENTAGE", "7").getValue());
        var max = Integer.parseInt(configurationBo.findOrSetDefault("ALLIANCE_MAX_SIZE", "15").getValue());
        float allowedByPercentage = userCount * (allowedPercentage / 100);
        allowedByPercentage = allowedByPercentage < 2 ? 2 : allowedByPercentage;
        int maxAllowed = (int) (allowedByPercentage < max ? allowedByPercentage : max);
        if (maxAllowed <= userStorageRepository.countByAlliance(request.getAlliance())) {
            throw new SgtBackendInvalidInputException("I18N_ERR_ALLIANCE_IS_FULL");
        }

    }

    /**
     * Defines the new alliance for all the users having and old alliance <br>
     * Usually used to delete an alliance
     *
     * @author Kevin Guanche Darias <kevin@kevinguanchedarias.com>
     * @since 0.7.0
     */
    private void defineAllianceByAllianceId(Integer oldAlliance) {
        userStorageRepository.defineAllianceByAllianceId(findById(oldAlliance), null);
    }
}

File: rust-backend/owge-business/src/bo/alliance_bo.rs
//! Port of `AllianceBo` + `AllianceJoinRequestBo` (the controller-reachable
//! paths only). The two Java Bos are folded into one Rust module: the Java
//! `AllianceJoinRequestBo.save` (with its duplicate-check) is dead code on the
//! `AllianceRestService` path — `requestJoin` inserts directly via the
//! repository — so only the reachable behaviour is ported.
//!
//! Auditing (`auditBo.doAudit` / `nonRequestAudit`) is dropped (not ported); the
//! live port disables it, so it is intentionally a no-op. Websocket emissions are M4;
//! the Java code does not emit on these paths, so there are none to defer here.

use crate::dto::{AllianceDto, AllianceJoinRequestDto, SimpleUserData};
use crate::error::{OwgeError, OwgeResult};
use crate::model::Alliance;
use sqlx::{Connection, MySqlConnection};

use super::ConfigurationBo;

pub struct AllianceBo;

/// `alliances` row joined with nothing — straight column mapping.
#[derive(sqlx::FromRow)]
struct AllianceRow {
    id: u16,
    name: String,
    description: Option<String>,
    image: Option<String>,
    owner_id: i32,
}

impl From<AllianceRow> for AllianceDto {
    fn from(r: AllianceRow) -> Self {
        AllianceDto {
            id: r.id,
            name: r.name,
            description: r.description,
            image: r.image,
            owner: r.owner_id,
        }
    }
}

impl From<Alliance> for AllianceDto {
    fn from(a: Alliance) -> Self {
        AllianceDto {
            id: a.id,
            name: a.name,
            description: a.description,
            image: a.image,
            owner: a.owner_id,
        }
    }
}

const ALLIANCE_COLS: &str = "id, name, description, image, owner_id";

impl AllianceBo {
    /// `AllianceRestService.findAll` -> `allianceRepository.findAll()`.
    pub async fn find_all(conn: &mut MySqlConnection) -> OwgeResult<Vec<AllianceDto>> {
        let rows = sqlx::query_as::<_, AllianceRow>(
            "SELECT id, name, description, image, owner_id FROM alliances ORDER BY id",
        )
        .fetch_all(&mut *conn)
        .await?;
        Ok(rows.into_iter().map(Into::into).collect())
    }

    async fn find_by_id_opt(conn: &mut MySqlConnection, id: u16) -> OwgeResult<Option<Alliance>> {
        let row = sqlx::query_as::<_, Alliance>(&format!(
            "SELECT {ALLIANCE_COLS} FROM alliances WHERE id = ?"
        ))
        .bind(id)
        .fetch_optional(&mut *conn)
        .await?;
        Ok(row)
    }

    async fn find_by_id_or_die(conn: &mut MySqlConnection, id: u16) -> OwgeResult<Alliance> {
        Self::find_by_id_opt(&mut *conn, id)
            .await?
            .ok_or_else(|| OwgeError::NotFound(format!("No alliance with id {id}")))
    }

    /// `AllianceRestService.members` -> `AllianceRepository.findMembers` — the
    /// users whose `alliance_id` is this alliance. Email is blanked and
    /// improvements omitted (the Rust DTO has no improvements field), matching
    /// the Java controller's `setEmail(null); setImprovements(null)`.
    pub async fn members(
        conn: &mut MySqlConnection,
        alliance_id: u16,
    ) -> OwgeResult<Vec<SimpleUserData>> {
        let rows = SimpleUserData::builder_select()
            .id(&4)?
            .alliance_id(&Some(alliance_id))?
            .find_all(&mut *conn)
            .await?;

        Ok(rows)
    }

    /// `AllianceBo.save(alliance, invokerId)` — create (id null) or update
    /// (id present). Runs in a transaction so the create path's two writes
    /// (insert alliance + point owner at it) are atomic.
    pub async fn save(
        conn: &mut MySqlConnection,
        dto: AllianceDto,
        invoker_id: i32,
    ) -> OwgeResult<AllianceDto> {
        let disabled =
            ConfigurationBo::find_or_set_default(&mut *conn, "DISABLED_FEATURE_ALLIANCE", "FALSE")
                .await?
                .value;
        if disabled.eq_ignore_ascii_case("true") {
            return Err(OwgeError::InvalidInput(
                "You can't not create an alliance, while the idea is nice, it's not possible"
                    .into(),
            ));
        }

        // On the Rust side the create path is distinguished by id == 0 (no body
        // id sent). The frontend omits/zeroes id when creating.
        let is_create = dto.id == 0;

        if is_create {
            let creator = SimpleUserData::find_by_id(&mut *conn, &invoker_id).await?;

            if creator.alliance_id.is_some() {
                return Err(OwgeError::InvalidInput(
                    "You already have an alliance, leave it first".into(),
                ));
            }

            let mut tx = conn.begin().await?;
            let result = sqlx::query(
                "INSERT INTO alliances (name, description, image, owner_id) VALUES (?, ?, ?, ?)",
            )
            .bind(&dto.name)
            .bind(&dto.description)
            .bind(&dto.image)
            .bind(invoker_id)
            .execute(&mut *tx)
            .await?;
            let new_id = result.last_insert_id() as u16;

            sqlx::query("UPDATE user_storage SET alliance_id = ? WHERE id = ?")
                .bind(new_id)
                .bind(invoker_id)
                .execute(&mut *tx)
                .await?;
            tx.commit().await?;
            // auditBo.doAudit(JOIN_ALLIANCE) — NOT PORTED (auditing disabled in the
            // live port and intentionally not ported.

            Ok(AllianceDto {
                id: new_id,
                name: dto.name,
                description: dto.description,
                image: dto.image,
                owner: invoker_id,
            })
        } else {
            let stored = Self::find_by_id_or_die(&mut *conn, dto.id).await?;
            Self::check_invoker_is_owner(&stored, invoker_id)?;
            // Update only name + description (matching Java).
            sqlx::query("UPDATE alliances SET name = ?, description = ? WHERE id = ?")
                .bind(&dto.name)
                .bind(&dto.description)
                .bind(stored.id)
                .execute(&mut *conn)
                .await?;
            Ok(AllianceDto {
                id: stored.id,
                name: dto.name,
                description: dto.description,
                image: stored.image,
                owner: stored.owner_id,
            })
        }
    }

    /// `AllianceBo.deleteByUser` — delete the invoker's own alliance (they must
    /// own one). Unsets every member's `alliance_id`, deletes the alliance's
    /// join requests, then deletes the alliance row.
    pub async fn delete_by_user(conn: &mut MySqlConnection, invoker_id: i32) -> OwgeResult<()> {
        let user = SimpleUserData::find_by_id(&mut *conn, &invoker_id).await?;

        let alliance_id = user
            .alliance_id
            .ok_or_else(|| OwgeError::InvalidInput("You don't have any alliance".into()))?;

        let alliance = Self::find_by_id_or_die(&mut *conn, alliance_id).await?;

        Self::check_invoker_is_owner(&alliance, invoker_id)?;

        let mut tx = conn.begin().await?;

        sqlx::query("UPDATE user_storage SET alliance_id = NULL WHERE alliance_id = ?")
            .bind(alliance_id)
            .execute(&mut *tx)
            .await?;
        sqlx::query("DELETE FROM alliance_join_request WHERE alliance_id = ?")
            .bind(alliance_id)
            .execute(&mut *tx)
            .await?;
        sqlx::query("DELETE FROM alliances WHERE id = ?")
            .bind(alliance_id)
            .execute(&mut *tx)
            .await?;
        tx.commit().await?;
        Ok(())
    }

    /// `AllianceRestService.listRequest` — join requests for the invoker's
    /// alliance. The invoker must own an alliance.
    pub async fn list_request(
        conn: &mut MySqlConnection,
        invoker_id: i32,
    ) -> OwgeResult<Vec<AllianceJoinRequestDto>> {
        let user = SimpleUserData::find_by_id(&mut *conn, &invoker_id).await?;

        let alliance_id = user.require_alliance()?;

        let alliance = Self::find_by_id_or_die(&mut *conn, alliance_id).await?;
        alliance.check_owner(invoker_id)?;

        Self::find_join_request_dtos(&mut *conn, "alliance_id", alliance_id as i64).await
    }

    /// `AllianceRestService.myRequests` -> `findByUserId`.
    pub async fn my_requests(
        conn: &mut MySqlConnection,
        invoker_id: i32,
    ) -> OwgeResult<Vec<AllianceJoinRequestDto>> {
        Self::find_join_request_dtos(&mut *conn, "user_id", invoker_id as i64).await
    }

    /// `AllianceRestService.myRequestsDelete` -> bare `deleteById` (no checks,
    /// matching Java).
    pub async fn delete_join_request_by_id(conn: &mut MySqlConnection, id: u32) -> OwgeResult<()> {
        sqlx::query("DELETE FROM alliance_join_request WHERE id = ?")
            .bind(id)
            .execute(&mut *conn)
            .await?;
        Ok(())
    }

    /// `AllianceBo.requestJoin` — create a join request. The Java path inserts
    /// directly via the repository (no duplicate check), so we replicate that
    /// and set `request_date = now (UTC)` since the column is NOT NULL.
    pub async fn request_join(
        conn: &mut MySqlConnection,
        alliance_id: u16,
        invoker_id: i32,
    ) -> OwgeResult<AllianceJoinRequestDto> {
        let alliance = Self::find_by_id_or_die(&mut *conn, alliance_id).await?;
        let user = SimpleUserData::find_by_id(&mut *conn, &invoker_id).await?;

        user.check_no_alliance()?;

        let now = chrono::Utc::now().naive_utc();
        let result = sqlx::query(
            "INSERT INTO alliance_join_request (alliance_id, user_id, request_date) VALUES (?, ?, ?)",
        )
        .bind(alliance_id)
        .bind(invoker_id)
        .bind(now)
        .execute(&mut *conn)
        .await?;
        let new_id = result.last_insert_id() as u32;
        Ok(AllianceJoinRequestDto {
            id: new_id,
            user,
            alliance: alliance.into(),
        })
    }

    /// `AllianceBo.acceptJoin` — accept a pending request: the invoker must own
    /// the request's alliance, the alliance must not be full, and if the
    /// requesting user has no alliance their `alliance_id` is set and all of
    /// their join requests are removed; otherwise just this request is removed.
    pub async fn accept_join(
        conn: &mut MySqlConnection,
        join_request_id: u32,
        invoker_id: i32,
    ) -> OwgeResult<()> {
        let request = Self::find_join_request_or_die(&mut *conn, join_request_id).await?;
        let alliance = Self::find_by_id_or_die(&mut *conn, request.alliance_id).await?;
        Self::check_invoker_is_owner(&alliance, invoker_id)?;
        Self::check_is_limit_reached(&mut *conn, request.alliance_id).await?;

        let requesting_user = SimpleUserData::find_by_id(&mut *conn, &request.user_id).await?;

        let mut tx = conn.begin().await?;

        if requesting_user.alliance_id.is_none() {
            sqlx::query("UPDATE user_storage SET alliance_id = ? WHERE id = ?")
                .bind(request.alliance_id)
                .bind(request.user_id)
                .execute(&mut *tx)
                .await?;

            sqlx::query("DELETE FROM alliance_join_request WHERE user_id = ?")
                .bind(request.user_id)
                .execute(&mut *tx)
                .await?;
        } else {
            // If accepts but the user already has an alliance, remove its join request
            sqlx::query("DELETE FROM alliance_join_request WHERE id = ?")
                .bind(join_request_id)
                .execute(&mut *tx)
                .await?;
        }
        tx.commit().await?;
        Ok(())
    }

    /// `AllianceBo.rejectJoin` — invoker must own the request's alliance; delete
    /// the request.
    pub async fn reject_join(
        conn: &mut MySqlConnection,
        join_request_id: u32,
        invoker_id: i32,
    ) -> OwgeResult<()> {
        let request = Self::find_join_request_or_die(&mut *conn, join_request_id).await?;
        let alliance = Self::find_by_id_or_die(&mut *conn, request.alliance_id).await?;
        Self::check_invoker_is_owner(&alliance, invoker_id)?;
        sqlx::query("DELETE FROM alliance_join_request WHERE id = ?")
            .bind(join_request_id)
            .execute(&mut *conn)
            .await?;
        Ok(())
    }

    /// `AllianceBo.leave` — owners can't leave their own alliance; otherwise
    /// unset the invoker's `alliance_id`.
    pub async fn leave(conn: &mut MySqlConnection, invoker_id: i32) -> OwgeResult<()> {
        let owns: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM alliances WHERE owner_id = ?")
            .bind(invoker_id)
            .fetch_one(&mut *conn)
            .await?;
        if owns > 0 {
            return Err(OwgeError::InvalidInput(
                "You can't leave your own alliance".into(),
            ));
        }
        sqlx::query("UPDATE user_storage SET alliance_id = NULL WHERE id = ?")
            .bind(invoker_id)
            .execute(&mut *conn)
            .await?;
        Ok(())
    }

    fn check_invoker_is_owner(alliance: &Alliance, invoker_id: i32) -> OwgeResult<()> {
        if alliance.owner_id != invoker_id {
            return Err(OwgeError::InvalidInput(
                "You are NOT the owner of that alliance, try hacking the owner account".into(),
            ));
        }
        Ok(())
    }

    /// `AllianceBo.checkIsLimitReached`.
    async fn check_is_limit_reached(
        conn: &mut MySqlConnection,
        alliance_id: u16,
    ) -> OwgeResult<()> {
        let user_count: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM user_storage")
            .fetch_one(&mut *conn)
            .await?;
        let allowed_percentage: i64 =
            ConfigurationBo::find_or_set_default(&mut *conn, "ALLIANCE_MAX_SIZE_PERCENTAGE", "7")
                .await?
                .value
                .trim()
                .parse()
                .unwrap_or(0);
        let max: i64 = ConfigurationBo::find_or_set_default(&mut *conn, "ALLIANCE_MAX_SIZE", "15")
            .await?
            .value
            .trim()
            .parse()
            .unwrap_or(0);
        let mut allowed_by_percentage = user_count as f32 * (allowed_percentage as f32 / 100.0);
        if allowed_by_percentage < 2.0 {
            allowed_by_percentage = 2.0;
        }
        let max_allowed = if allowed_by_percentage < max as f32 {
            allowed_by_percentage as i64
        } else {
            max
        };
        let in_alliance: i64 =
            sqlx::query_scalar("SELECT COUNT(*) FROM user_storage WHERE alliance_id = ?")
                .bind(alliance_id)
                .fetch_one(&mut *conn)
                .await?;
        if max_allowed <= in_alliance {
            return Err(OwgeError::InvalidInput("I18N_ERR_ALLIANCE_IS_FULL".into()));
        }
        Ok(())
    }

    async fn find_join_request_or_die(
        conn: &mut MySqlConnection,
        id: u32,
    ) -> OwgeResult<crate::model::AllianceJoinRequest> {
        let row = sqlx::query_as::<_, crate::model::AllianceJoinRequest>(
            "SELECT id, alliance_id, user_id, request_date FROM alliance_join_request WHERE id = ?",
        )
        .bind(id)
        .fetch_optional(&mut *conn)
        .await?;
        row.ok_or_else(|| OwgeError::NotFound(format!("No join request with id {id}")))
    }

    /// Loads join requests filtered on `column = value` (column is a fixed
    /// internal literal, never user input), eager-joining the user and alliance
    /// into full DTOs.
    async fn find_join_request_dtos(
        conn: &mut MySqlConnection,
        column: &str,
        value: i64,
    ) -> OwgeResult<Vec<AllianceJoinRequestDto>> {
        let requests = sqlx::query_as::<_, crate::model::AllianceJoinRequest>(&format!(
            "SELECT id, alliance_id, user_id, request_date FROM alliance_join_request \
             WHERE {column} = ? ORDER BY id"
        ))
        .bind(value)
        .fetch_all(&mut *conn)
        .await?;

        let mut out = Vec::with_capacity(requests.len());
        for req in requests {
            let user = SimpleUserData::find_by_id(&mut *conn, &req.user_id).await?;

            let alliance = Self::find_by_id_or_die(&mut *conn, req.alliance_id).await?;
            out.push(AllianceJoinRequestDto {
                id: req.id,
                user,
                alliance: alliance.into(),
            });
        }
        Ok(out)
    }
}
